include/engine/guidance/api_response_generator.hpp

```diff
@@ -0,0 +1,286 @@
+    osrm::json::Array ListViaPoints(const InternalRouteResult &raw_route) const;
+    osrm::json::Array ListViaIndices(const SegmentListT &segment_list) const;
+
+    osrm::json::Value GetGeometry(const bool return_encoded, const SegmentListT &segments) const;
+
+    // TODO this dedicated creation seems unnecessary? Only used for route names
+    std::vector<Segment> BuildRouteSegments(const SegmentListT &segment_list) const;
+
+    // adds checksum and locations
+    osrm::json::Object BuildHintData(const InternalRouteResult &raw_route) const;
+
+  private:
+    // data access to translate ids back into names
+    DataFacade *facade;
+};
+
+template <typename DataFacadeT>
+ApiResponseGenerator<DataFacadeT>::ApiResponseGenerator(DataFacadeT *facade_)
+    : facade(facade_)
+{
+}
+
+template <typename DataFacadeT>
+void ApiResponseGenerator<DataFacadeT>::DescribeRoute(const DescriptorConfig &config,
+                                                      const InternalRouteResult &raw_route,
+                                                      osrm::json::Object &json_result)
+{
+    if( not raw_route.is_valid() ){
+      return;
+    }
+    const bool ALLOW_SIMPLIFICATION = true;
+    const bool EXTRACT_ROUTE = false;
+    const bool EXTRACT_ALTERNATIVE = true;
+    SegmentListT segment_list(raw_route, EXTRACT_ROUTE, config.zoom_level, ALLOW_SIMPLIFICATION,
+                              facade);
+    json_result.values["route_summary"] = SummarizeRoute(raw_route, segment_list);
+    json_result.values["via_points"] = ListViaPoints(raw_route);
+    json_result.values["via_indices"] = ListViaIndices(segment_list);
+
+    if (config.geometry)
+    {
+        json_result.values["route_geometry"] = GetGeometry(config.encode_geometry, segment_list);
+    }
+
+    if (config.instructions)
+    {
+        json_result.values["route_instructions"] =
+            annotators::AnnotateRoute(segment_list.Get(), facade);
+    }
+
+    RouteNames route_names;
+    RouteNameExtractor generate_route_names;
+
+    if (raw_route.has_alternative())
+    {
+        SegmentListT alternate_segment_list(raw_route, EXTRACT_ALTERNATIVE, config.zoom_level,
+                                            ALLOW_SIMPLIFICATION, facade);
+
+        // Alternative Route Summaries are stored in an array to (down the line) allow multiple
+        // alternatives
+        osrm::json::Array json_alternate_route_summary_array;
+        json_alternate_route_summary_array.values.push_back(
+            SummarizeRoute(raw_route, alternate_segment_list));
+        json_result.values["alternative_summaries"] = json_alternate_route_summary_array;
+        json_result.values["alternative_indices"] = ListViaIndices(alternate_segment_list);
+
+        if (config.geometry)
+        {
+            osrm::json::Value alternate_geometry_string =
+                GetGeometry(config.encode_geometry, alternate_segment_list);
```

Comment:
Hmm what I only now realize: I think you can `auto` away a lot of types here and throughout the codebase